### PR TITLE
fix: Fix example configs and README.md file

### PR DESCRIPTION
**Pull request by onidoru**
**What type of PR is this?**
documentation

**Which issue does this PR fix**:
Example LDAP configs do not represent the actual state of things and are outdated. 

```bash
# could also run $ make verify-config
# it's also the same problem with `examples/config-ldap-credentials.json`, so it's excluded from the Makefile
$ ./bin/zot-linux-amd64 verify examples/config-example.json 
{"level":"error","error":"1 error(s) decoding:\n\n* 'HTTP.Auth.LDAP' has invalid keys: binddn, bindpassword","time":"2024-01-16T23:36:27+02:00","message":"failed to unmarshaling new config"}
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by andaaron — COMMENTED** [review]
Change looks good, but we should make sure we also call `make verify-config` in CI.

File: pkg/cli/server/root.go [hunk 1]
```diff
@@ -861,8 +861,8 @@ func readLDAPCredentials(ldapConfigPath string) (config.LDAPCredentials, error)
 
 	var ldapCredentials config.LDAPCredentials
 
-	if err := viperInstance.Unmarshal(&ldapCredentials); err != nil {
-		log.Error().Err(err).Msg("failed to unmarshal new config")
+	if err := viperInstance.UnmarshalExact(&ldapCredentials); err != nil {
+		log.Error().Err(err).Msg("failed to unmarshal ldap credentials config")
```
Comment (andaaron) [inline pkg/cli/server/root.go:865]: Please add a test to cover this line. It should be easier than before, since we should fail in case of invalid or missing fields.

File: .github/workflows/verify-config.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-mod-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-mod-
+      - name: Install go dependencies
+        if: steps.cache-go-dependencies.outputs.cache-hit != 'true'
+        run: |
+          cd $GITHUB_WORKSPACE
+          go mod download
+      - uses: ./.github/actions/setup-localstack
```
Comment (andaaron) [inline .github/workflows/verify-config.yaml:42]: This should not be needed, as we don't test the S3 APIs in this workflow.

Comment (onidoru) [reply .github/workflows/verify-config.yaml]: Fixed.

File: pkg/cli/server/root.go (hunk 1 above)
Comment (onidoru) [reply pkg/cli/server/root.go:866]: Added new tests to ensure that credentials file contains all required fields or has any extra fields.

Please note that it does not check if [credentialsFile](https://github.com/project-zot/zot/pull/2215/files#diff-464834167072cf556f61bd7ba8a129ccde65577dc94ac2b0ce0ddecfb21d864cR21) field is present in `auth` config. Let me know if we want to change this old behavior and make it required.

File: pkg/cli/server/root.go [hunk 3]
```diff
@@ -861,12 +861,34 @@ func readLDAPCredentials(ldapConfigPath string) (config.LDAPCredentials, error)
...
-		log.Error().Err(err).Msg("failed to unmarshal new config")
+	metaData := &mapstructure.Metadata{}
+	if err := viperInstance.UnmarshalExact(&ldapCredentials, metadataConfig(metaData)); err != nil {
+		log.Error().Err(err).Msg("failed to unmarshal ldap credentials config")
 
 		return config.LDAPCredentials{}, err
 	}
 
+	if len(metaData.Keys) == 0 {
+		log.Error().Err(zerr.ErrBadConfig).
```
Comment (andaaron) [inline pkg/cli/server/root.go:872]: Can you cover this case? Should be easy enough.

File: pkg/cli/server/root.go [hunk 4]
```diff
@@ -861,12 +861,34 @@ func readLDAPCredentials(ldapConfigPath string) (config.LDAPCredentials, error)
...
+			Msg("failed to load ldap credentials config due to the absence of any key:value pair")
+
+		return config.LDAPCredentials{}, zerr.ErrBadConfig
+	}
+
+	if len(metaData.Unused) > 0 {
```
Comment (andaaron) [inline pkg/cli/server/root.go:878]: Can you cover this with a test case?

File: pkg/cli/server/root.go (hunk 3 above)
Comment (onidoru) [reply pkg/cli/server/root.go:873]: Covered.

File: pkg/cli/server/root.go (hunk 4 above)
Comment (onidoru) [reply pkg/cli/server/root.go:879]: Covered.

File: pkg/cli/server/root_test.go [hunk 5]
```diff
@@ -1236,6 +1236,161 @@ storage:
...
+
+		tmpCredsFile, err := os.CreateTemp("", "zot-cred*.json")
+		So(err, ShouldBeNil)
+		defer os.Remove(tmpCredsFile.Name())
+
+		// `bindDN` key is missing
+		content := []byte(`{
+    		"bindDN":"cn=ldap-searcher,ou=Users,dc=example,dc=org",
+    		"bindPassword":"ldap-searcher-password",
+			"extraKey": "extraValue"
```
Comment (andaaron) [inline pkg/cli/server/root_test.go:1330]: For some reason the test does not cover the lines we expected. Also the comment at line 1326 should probably be removed,

File: pkg/cli/server/root_test.go [hunk 6]
```diff
@@ -1236,6 +1236,161 @@ storage:
...
+		tmpFile, err := os.CreateTemp("", "zot-test*.json")
+		So(err, ShouldBeNil)
+		defer os.Remove(tmpFile.Name())
+
+		tmpCredsFile, err := os.CreateTemp("", "zot-cred*.json")
+		So(err, ShouldBeNil)
+		defer os.Remove(tmpCredsFile.Name())
+
+		// `bindDN` key is missing
+		content := []byte(``)
```
Comment (andaaron) [inline pkg/cli/server/root_test.go:1368]: I think this case covers the error on `viperInstance.UnmarshalExact(` which is good, but we still need a separate test case with a valid empty json `content := []byte(`{}`)`.

**Comment by rchincha** [comment]
Superseded by https://github.com/project-zot/zot/pull/2322
Thanks @onidoru for the PR, closing.
